### PR TITLE
pAI expansion

### PR DESCRIPTION
**Pull request by Dejaku51**
## About The Pull Request
PR expanding pAI containing fixes, new features etc
- [x] https://github.com/tgstation/tgstation/pull/62916 ~~[Requires TGUI Update?]~~
- [x] https://github.com/tgstation/tgstation/pull/63160 ~~[Requires TGUI Update?]~~
- [x] https://github.com/tgstation/tgstation/pull/56404
- [x] https://github.com/tgstation/tgstation/pull/52879
- [x] ~~https://github.com/tgstation/tgstation/pull/3142~~ [Abandoned]
- [x] https://github.com/tgstation/tgstation/pull/49684
- [x] https://github.com/tgstation/tgstation/pull/62462
- [x] https://github.com/tgstation/tgstation/pull/52575
- [x] https://github.com/tgstation/tgstation/pull/57033
- [x] https://github.com/tgstation/tgstation/pull/55215
- [x] https://github.com/tgstation/tgstation/pull/61112
- [x] https://github.com/tgstation/tgstation/pull/70869 [partly]

## Changelog
:cl:
add: Added new TGUI interface for pAIs
add: pAIs have now new software: GPS
add: pAIs have now new software: Photo printer
add: Borgs now can wear drones and pAIs as hats
add: pAIs now can use binary encryption keys
add: pAIs can now select holochassis using radial menu 
tweak: pAIs can now select radio channel from encryption keys
code: Cleaned up pAIs and aicard code
/:cl:

### CODE REVIEW

**Comment by Dejaku51** [comment]
fuck forgot to delete those

**Comment by Dejaku51** [comment]
I'm not gonna wait a month for someone to review it or smth. Feel free to create that PR yourself

**Comment by Dejaku51** [comment]
Fuck it, im just gonna leave it open. If any maintainer gonna review it please ping me on discord (Dejaku51#9677) so i can make changes

File: code/controllers/subsystem/pai.dm [hunk 1]
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
...
+		return
+	if(request_spam)
+		to_chat(user, "<span class='warning'>Request sent too recently.</span>")
+		return
+	request_spam = TRUE
+	playsound(src, 'sound/machines/ping.ogg', 20, TRUE)
+	to_chat(user, "<span class='notice'>You have requested pAI assistance.</span>")
+	var/mutable_appearance/alert_overlay = mutable_appearance('icons/obj/aicards.dmi', "pai")
+	notify_ghosts("[user] is requesting a pAI personality! Use the pAI button to submit yourself as one.", source=user, alert_overlay = alert_overlay, action=NOTIFY_ORBIT, header="pAI Request!", ignore_key = POLL_IGNORE_PAI)
+	addtimer(CALLBACK(src, .proc/request_again), 10 SECONDS)
```
Comment (ike709) [inline code/controllers/subsystem/pai.dm]: Why isn't this a `VARSET_CALLBACK()`?

File: code/controllers/subsystem/pai.dm [hunk 2]
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
...
+/datum/controller/subsystem/pai/proc/submit_alert()
+	if(submit_spam)
+		to_chat(usr, "<span class='warning'>Your candidacy has been submitted, but pAI cards have been alerted too recently.</span>")
+		return FALSE
+	submit_spam = TRUE
+	for(var/obj/item/paicard/paicard in pai_card_list)
+		if(!paicard.pai)
+			paicard.alertUpdate()
+	to_chat(usr, "<span class='notice'>Your pAI candidacy has been submitted!</span>")
+	addtimer(CALLBACK(src, .proc/submit_again), 10 SECONDS)
```
Comment (ike709) [inline code/controllers/subsystem/pai.dm]: Why isn't this a `VARSET_CALLBACK()`?

File: code/__DEFINES/pai.dm [hunk 3]
```diff
@@ -0,0 +1,2 @@
+/// Total percent required to hack a door.
+#define HACK_COMPLETE 100
```
Comment (ike709) [inline code/__DEFINES/pai.dm]: Including a new file for one define is kind of bleh IMO.

File: code/controllers/subsystem/pai.dm [hunk 4]
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
...
 	flags = SS_NO_INIT|SS_NO_FIRE
 
+	/// List of pAI candidates, including those not submitted.
 	var/list/candidates = list()
-	var/ghost_spam = FALSE
-	var/spam_delay = 100
+	/// Prevents a crew member from hitting "request pAI"
+	var/request_spam = FALSE
+	/// Prevents a pAI from submitting itself repeatedly and sounding an alert.
+	var/submit_spam = FALSE
```
Comment (ike709) [inline code/controllers/subsystem/pai.dm]: These are subsystem vars. Why does one person submitting a request once cause everyone to be ratelimited?

File: code/controllers/subsystem/pai.dm [hunk 5]
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
...
-/datum/controller/subsystem/pai/proc/recruitWindow(mob/M)
-	var/datum/paiCandidate/candidate
-	for(var/datum/paiCandidate/c in candidates)
-		if(c.key == M.key)
-			candidate = c
-	if(!candidate)
-		candidate = new /datum/paiCandidate()
-		candidate.key = M.key
+/// Created when a user clicks the "pAI candidate" window
+/datum/pai_candidate
```
Comment (ike709) [inline code/controllers/subsystem/pai.dm]: I'm not sure if these are filtered, but all of them (minus `key` & `ready`) need to be chat filtered when they're inputted.

File: code/controllers/subsystem/pai.dm [hunk 6]
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
...
+ * Finds the candidate in question from the list of candidates.
+ */
+/datum/controller/subsystem/pai/proc/check_candidate(mob/user)
+	/// Finds a matching candidate.
+	var/datum/pai_candidate/candidate
+	for(var/datum/pai_candidate/checked_candidate as anything in candidates)
+		if(checked_candidate.key == user.key)
+			candidate = checked_candidate
+			return candidate
+	return null
```
Comment (ike709) [inline code/controllers/subsystem/pai.dm]: Redundant.

File: code/controllers/subsystem/pai.dm [hunk 7]
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
...
+/**
+ * Finds the candidate in question from the list of candidates.
+ */
+/datum/controller/subsystem/pai/proc/check_candidate(mob/user)
+	/// Finds a matching candidate.
+	var/datum/pai_candidate/candidate
+	for(var/datum/pai_candidate/checked_candidate as anything in candidates)
+		if(checked_candidate.key == user.key)
+			candidate = checked_candidate
+			return candidate
```
Comment (ike709) [inline code/controllers/subsystem/pai.dm]: Why not just make `candidates` an assoc list of key & datum?

File: code/controllers/subsystem/pai.dm [hunk 8]
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
...
+		to_chat(user, "<span class='warning'>Due to growing incidents of SELF corrupted independent artificial intelligences, freeform personality devices have been temporarily banned in this sector.</span>")
+		return
+	if(request_spam)
+		to_chat(user, "<span class='warning'>Request sent too recently.</span>")
+		return
+	request_spam = TRUE
+	playsound(src, 'sound/machines/ping.ogg', 20, TRUE)
+	to_chat(user, "<span class='notice'>You have requested pAI assistance.</span>")
+	var/mutable_appearance/alert_overlay = mutable_appearance('icons/obj/aicards.dmi', "pai")
+	notify_ghosts("[user] is requesting a pAI personality! Use the pAI button to submit yourself as one.", source=user, alert_overlay = alert_overlay, action=NOTIFY_ORBIT, header="pAI Request!", ignore_key = POLL_IGNORE_PAI)
```
Comment (ike709) [inline code/controllers/subsystem/pai.dm:34]: Does this check pAI bans?

File: code/controllers/subsystem/pai.dm [hunk 9]
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
...
-		candidate = new /datum/paiCandidate()
-		candidate.key = M.key
+/// Created when a user clicks the "pAI candidate" window
+/datum/pai_candidate
+	/// User inputted OOC comments
+	var/comments
+	/// User inputted behavior description
+	var/description
+	/// User's ckey - not input
+	var/key
```
Comment (ike709) [inline code/controllers/subsystem/pai.dm]: This comment is a lie, you're using their `key` everywhere when it really should be `ckey` everywhere except `IS_GUEST_KEY()`.

File: code/modules/photography/camera/silicon_camera.dm [hunk 10]
```diff
@@ -98,5 +98,13 @@
 	p.pixel_x = rand(-10, 10)
 	p.pixel_y = rand(-10, 10)
 	C.toner -= printcost	 //All fun allowed.
-	visible_message("[C.name] spits out a photograph from a narrow slot on its chassis.")
-	to_chat(usr, "<span class='notice'>You print a photograph.</span>")
+	visible_message("[C.name] spits out a photograph from a narrow slot on its chassis.", "<span class='notice'>You print a photograph.</span>")
+
+/obj/item/camera/siliconcam/proc/paiprint(mob/user)
```
Comment (ike709) [inline code/modules/photography/camera/silicon_camera.dm:103]: Unused?

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 11]
```diff
@@ -1094,24 +1096,24 @@ GLOBAL_LIST_EMPTY(PDAs)
...
+	if (!choice)
 		return
 
-	var/selected = plist[c]
+	var/selected = plist[choice]
 
 	if(aicamera.stored.len)
-		var/add_photo = input(user,"Do you want to attach a photo?","Photo","No") as null|anything in list("Yes","No")
-		if(add_photo=="Yes")
+		var/add_photo = input(user,"Do you want to attach a photo?","Photo","No") as null|anything in list("Yes","No") /// Change to TGUI List once ported
```
Comment (ike709) [inline code/game/objects/items/devices/PDA/PDA.dm]: No.

File: code/game/objects/items/devices/gps.dm [hunk 12]
```diff
@@ -11,7 +11,10 @@
 
 /obj/item/gps/Initialize()
 	. = ..()
-	AddComponent(/datum/component/gps/item, gpstag)
+	add_gps_component()
+
+/// Adds the GPS component to this item.
+/obj/item/gps/proc/add_gps_component()
```
Comment (ike709) [inline code/game/objects/items/devices/gps.dm:17]:
```suggestion
/obj/item/gps/proc/add_gps_component()
	return
```

File: code/game/objects/items/devices/paicard.dm [hunk 13]
```diff
@@ -16,156 +21,167 @@
...
+	if(.)
+		return FALSE
+	switch(action)
+		if("download")
+			/// The individual candidate to download
+			var/datum/pai_candidate/candidate
+			for(var/datum/pai_candidate/checked_candidate as anything in SSpai.candidates)
+				if(params["key"] == checked_candidate.key)
+					candidate = checked_candidate
+					break
```
Comment (ike709) [inline code/game/objects/items/devices/paicard.dm]: Look another place where it being assoc would simplify things.

**Comment by Dejaku51** [comment]
oh god there is more

File: code/modules/mob/living/silicon/pai/pai.dm [hunk 14]
```diff
@@ -95,8 +111,33 @@
 /mob/living/silicon/pai/can_buckle()
 	return FALSE
 
+/mob/living/silicon/pai/handle_atom_del(atom/A)
+	if(A == hacking_cable)
```
Comment (ike709) [inline code/modules/mob/living/silicon/pai/pai.dm:117]: Couldn't this all be a switch()?

File: code/controllers/subsystem/pai.dm (hunk 9 above)
Comment (Dejaku51) [reply code/controllers/subsystem/pai.dm]: so should i change `key` in comment to `ckey` or change every `key` in my code to `ckey` ?

**Review by ike709 — CHANGES_REQUESTED** [review]
Gave this a once-over from ~2-6am whenever I was waiting on something to happen in Minecraft.

Looked at everything except the software file.

Also don't just manually resolve reviews without leaving a comment explaining why it's resolved.

File: code/controllers/subsystem/pai.dm (hunk 9 above)
Comment (ike709) [reply code/controllers/subsystem/pai.dm]: Latter, minus the guest key check.

**Comment by Dejaku51** [comment]
> Gave this a once-over from ~2-6am whenever I was waiting on something to happen in Minecraft.
> 
> Looked at everything except the software file.
> 
> Also don't just manually resolve reviews without leaving a comment explaining why it's resolved.

i just fixed these things that i marked as resolved

File: code/__DEFINES/misc.dm [hunk 15]
```diff
@@ -467,3 +467,7 @@ GLOBAL_LIST_INIT(pda_styles, sortList(list(MONO, VT, ORBITRON, SHARE)))
 #define ALIGNMENT_GOOD "good"
 #define ALIGNMENT_NEUT "neutral"
 #define ALIGNMENT_EVIL "evil"
+
+// pAI
+
+#define HACK_COMPLETE 100 // Total percent required to hack a door.
```
Comment (ike709) [inline code/__DEFINES/misc.dm]: Don't make a define for 100?

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 11 above)
Comment (Archanial) [reply code/game/objects/items/devices/PDA/PDA.dm]: (ike meant comment, we're not porting tgui lists)

**Comment by ike709** [comment]
Unaddressed reviews, fails CI

File: code/modules/mob/living/silicon/pai/pai.dm (hunk 14 above)
Comment (Dejaku51) [reply code/modules/mob/living/silicon/pai/pai.dm:117]: don't think so, switch requires constant expression

**Comment by Dejaku51** [comment]
fuck you github actions

**Comment by PowerfulBacon** [comment]
Dchat broke, could be related.
Runtime in paicard.dm, line 204: Cannot read [ckey].ready 
[09:29:45] Runtime in mobs.dm, line 434: Cannot read null.prefs
proc name: deadchat broadcast (/proc/deadchat_broadcast)

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]
Breaks GPSs, provided runtimes above and dchat no longer works (is last one related?)

**Comment by PowerfulBacon** [comment]
PAI candidate loading doesn't work

**Comment by jlsnow301** [comment]
issue with candidate state is still present here, check this for fix
https://github.com/tgstation/tgstation/pull/65071

i am also saddened that you removed the pai's cat wallpaper

**Comment by PowerfulBacon** [comment]
Fix candidate issue and this PR will be next in line for testmerge.

**Comment by Dejaku51** [comment]
> i am also saddened that you removed the pai's cat wallpaper

i liked it but removal was requested by people on discord

**Comment by Dejaku51** [comment]
> issue with candidate state is still present here, check this for fix
> https://github.com/tgstation/tgstation/pull/65071

i thought i did that huh

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]
pAI laws should be stated in the chat when they change.
Display image button doesn't work
ui_act should return true, the UI is very unresponsive when updating.
When in holoform folded out, I randomly take damage constantly.
image
pAIs also have no way to leave the game themselves and other players cannot delete their pAIs now apparently.

**Review by PowerfulBacon — DISMISSED** [review]
Suiciding as a pAI doesnt remove you from the list of pAIs that can be registered (Does it do this if you are downloaded already)

**Comment by Dejaku51** [comment]
> Suiciding as a pAI doesnt remove you from the list of pAIs that can be registered (Does it do this if you are downloaded already)

Fixed

> pAI laws should be stated in the chat when they change.

Should work now

> Display image button doesn't work

Fixed

> ui_act should return true, the UI is very unresponsive when updating.

Done

> When in holoform folded out, I randomly take damage constantly.

Couldn't reproduce, probably something was happening during that particular round

> pAIs also have no way to leave the game themselves and other players cannot delete their pAIs now apparently.

Added wipe button

**Comment by FortSumter** [comment]
I like this a lot. One thing I've wanted as a PAI was to vent crawl as a hologram. Mechanically this shouldn't be a problem as a PAI can't mess with anything on it's own bit it would be able to spy on people for antags or security while also not getting stuck behind doors when wandering.

I'd also like to be able to jump into people's hands back again. It was such a wholesome feature!

**Comment by Dejaku51** [comment]
@itsmeow i have currently some personal projects so i don't have time to do it, but i'll do it once im able to

**Comment by Dejaku51** [comment]
fixed but we should TM it to make sure

**Comment by Dejaku51** [comment]
check if owner is still in range when changing laws

**Comment by Dejaku51** [comment]
DNA verify

**Comment by Dejaku51** [comment]
capitalize software names

**Comment by Dejaku51** [comment]
pAIs have all huds

**Comment by Dejaku51** [comment]
pAIs start with messanger

**Comment by Dejaku51** [comment]
change buttons in host scan

**Comment by Dejaku51** [comment]
can't interact with internal radio

**Comment by itsmeow** [comment]
image

**Comment by Dejaku51** [comment]
runtime: Progressbar.dm, line 61 Bad index

**Comment by Dejaku51** [comment]
force update on submit

**Comment by benbot16** [comment]
PAI hud doesn't really scale, dunno if that's really a bug though.

image

**Comment by Dejaku51** [comment]
make reemaggable

**Comment by benbot16** [comment]
Subspace transceiver doesn't work if you're not in holoform mode (or if you're picked up).

**Comment by benbot16** [comment]
PAIs can be carried without an arm.

**Comment by itsmeow** [comment]
how does one even discover this, insanity

**Comment by benbot16** [comment]
it might be delimber fuckery, don't know exactly though.

**Comment by Dejaku51** [comment]
most likely delimber

**Comment by benbot16** [comment]
Confirmed as delimber fuckery.

**Comment by Dejaku51** [comment]
going to sleep now so i'll test it tomorrow

**Comment by Dejaku51** [comment]
amp keys work on pAIs

File: code/modules/mob/living/silicon/silicon.dm [hunk 16]
```diff
@@ -24,7 +24,7 @@
 	var/obj/item/camera/siliconcam/aicamera = null //photography
 	hud_possible = list(ANTAG_HUD, DIAG_STAT_HUD, DIAG_HUD, DIAG_TRACK_HUD)
 
-	var/obj/item/radio/borg/radio = null //All silicons make use of this, with (p)AI's creating headsets
+	var/obj/item/radio/borg/pai/radio = null //All silicons make use of this, with (p)AI's creating headsets
```
Comment (itsmeow) [inline code/modules/mob/living/silicon/silicon.dm]: This is for all silicons
you can assign a pAI radio to it but don't change the base type

File: code/modules/mob/living/silicon/pai/software.dm [hunk 17]
```diff
@@ -206,7 +204,7 @@
 	if(!istype(master))
 		return
 	to_chat(pai, "<span class='notice'>Requesting a DNA sample.</span>")
-	var/confirm = alert(master, "[pai] is requesting a DNA sample from you. Will you allow it to confirm your identity?", "Checking DNA", list("Yes", "No"))
+	var/confirm = input(master, "[pai] is requesting a DNA sample from you. Will you allow it to confirm your identity?", "Checking DNA") as anything in list("Yes", "No")
```
Comment (itsmeow) [inline code/modules/mob/living/silicon/pai/software.dm]: this will not work

File: code/modules/mob/living/silicon/pai/software.dm [hunk 18]
```diff
@@ -4,24 +4,24 @@
...
-															"door jack" = 25,
-															"encryption keys" = 25,
-															"internal gps" = 35,
-															"universal translator" = 35
+															"Crew Manifest" = 5,
+															"Digital Messenger" = 5,
+															"Atmosphere Sensor" = 5,
+															"Photography Module" = 5,
+															"Camera Zoom" = 10,
+															"Printer Module" = 10,
```
Comment (itsmeow) [inline code/modules/mob/living/silicon/pai/software.dm]: make all of the software names into DEFINEs please

File: code/game/objects/items/devices/radio/radio.dm [hunk 19]
```diff
@@ -456,3 +456,9 @@
 /obj/item/radio/off	// Station bounced radios, their only difference is spawning with the speakers off, this was made to help the lag.
 	listening = 0			// And it's nice to have a subtype too for future features.
 	dog_fashion = /datum/dog_fashion/back
+
+/obj/item/radio/borg/pai
+	name = "pai radio"
+/obj/item/radio/borg/pai/ui_state(mob/user)
```
Comment (itsmeow) [inline code/game/objects/items/devices/radio/radio.dm]: spacing between these two lines please

File: code/game/objects/items/devices/paicard.dm [hunk 20]
```diff
@@ -132,6 +134,8 @@
 				pai.laws.set_zeroth_law("Serve your master.")
 				pai.emittersemicd = FALSE
 		if("set_laws")
+			if(!in_range(src, usr))
```
Comment (itsmeow) [inline code/game/objects/items/devices/paicard.dm:137]: this needs to go AFTER the stripped_multiline_input

File: code/game/objects/items/devices/paicard.dm [hunk 21]
```diff
@@ -14,6 +14,7 @@
 	///what emotion icon we have. handled in /mob/living/silicon/pai/Topic()
 	var/emotion_icon = "off"
 	resistance_flags = FIRE_PROOF | ACID_PROOF | INDESTRUCTIBLE
+	emag_toggleable = TRUE
```
Comment (itsmeow) [inline code/game/objects/items/devices/paicard.dm]: don't do this, override can_emag and return true like the sleeper does

File: tgui/packages/tgui/interfaces/PaiInterface.tsx [hunk 22]
```diff
@@ -71,36 +71,36 @@ seen as the ideal, unwavering human companion that you are.`;
...
     'A tool that allows you to broadcast news to other crew members.',
-  'door jack': 'A tool that allows you to open doors.',
-  'encryption keys':
+  'Door Jack': 'A tool that allows you to open doors.',
+  'Encryption Keys':
     'A tool that allows you to decrypt and speak on other radio frequencies.',
-  'internal gps': 'A tool that allows you to track your location.',
-  'universal translator': 'Translation module for non-common languages.',
+  'Internal GPS': 'A tool that allows you to track your location.',
+  'Universal Translator': 'Translation module for non-common languages.',
```
Comment (itsmeow) [inline tgui/packages/tgui/interfaces/PaiInterface.tsx:97]: I hate that this isn't a field on the software types passed in through static data

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/silicon/pai/software.dm (hunk 18 above)
Comment (Dejaku51) [reply code/modules/mob/living/silicon/pai/software.dm]: done

File: code/game/objects/items/devices/radio/radio.dm (hunk 19 above)
Comment (Dejaku51) [reply code/game/objects/items/devices/radio/radio.dm]: removed entirely

File: code/game/objects/items/devices/paicard.dm (hunk 20 above)
Comment (Dejaku51) [reply code/game/objects/items/devices/paicard.dm:137]: done

File: code/game/objects/items/devices/paicard.dm (hunk 21 above)
Comment (Dejaku51) [reply code/game/objects/items/devices/paicard.dm]: done (at least i think i did it correctly)

**Comment by Dejaku51** [comment]
oh fuck you

File: code/modules/mob/living/silicon/pai/software.dm (hunk 17 above)
Comment (Dejaku51) [reply code/modules/mob/living/silicon/pai/software.dm]: now?

File: tgui/packages/tgui/interfaces/PaiInterface.tsx (hunk 22 above)
Comment (Dejaku51) [reply tgui/packages/tgui/interfaces/PaiInterface.tsx:97]: is it a request or just complaining

File: code/modules/mob/living/silicon/silicon.dm (hunk 16 above)
Comment (Dejaku51) [reply code/modules/mob/living/silicon/silicon.dm]: fixed

File: tgui/packages/tgui/interfaces/PaiInterface.tsx (hunk 22 above)
Comment (itsmeow) [reply tgui/packages/tgui/interfaces/PaiInterface.tsx:97]: It is a request phrased as complaining

File: code/modules/mob/living/silicon/pai/pai.dm [hunk 23]
```diff
@@ -375,7 +373,7 @@
 	return ..()
 
 /obj/item/paicard/should_emag(mob/user)
-	return pai && ..()
+	return TRUE
```
Comment (itsmeow) [inline code/modules/mob/living/silicon/pai/pai.dm]: return `!!pai`, my bad

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/silicon/pai/software.dm (hunk 17 above)
Comment (itsmeow) [reply code/modules/mob/living/silicon/pai/software.dm]: Looks correct

File: tgui/packages/tgui/interfaces/PaiInterface.tsx (hunk 22 above)
Comment (itsmeow) [reply tgui/packages/tgui/interfaces/PaiInterface.tsx:97]: It's also optional though

File: code/modules/mob/living/silicon/pai/pai.dm (hunk 23 above)
Comment (Dejaku51) [reply code/modules/mob/living/silicon/pai/pai.dm]: done

**Comment by Dejaku51** [comment]
guh?

**Comment by Dejaku51** [comment]
i'll fix it on 30 december :trollface:

**Comment by ike709** [comment]
Fails CI

**Comment by Tsar-Salat** [comment]
The day approaches...

**Review by itsmeow — APPROVED** [review]
1 year. Amazing.

**Review by ike709 — APPROVED** [review]
YOLO

**Pull request merged**